connection/cluster/ClusterServerStub.java

```diff
@@ -85,30 +107,62 @@ public static ClusterServerStub create(String username, String password, Managed
         return resilientCall(() -> clusterBlockingStub.databasesAll(request));
     }
 
-    public static class CredentialEmbedder extends CallCredentials {
-        private static final Metadata.Key<String> USERNAME_FIELD = Metadata.Key.of("username", ASCII_STRING_MARSHALLER);
-        private static final Metadata.Key<String> PASSWORD_FIELD = Metadata.Key.of("password", ASCII_STRING_MARSHALLER);
+    @Override
+    protected ManagedChannel channel() {
+        return channel;
+    }
 
-        private final String username;
-        private final String password;
+    @Override
+    protected TypeDBGrpc.TypeDBBlockingStub blockingStub() {
+        return blockingStub;
+    }
+
+    @Override
+    protected TypeDBGrpc.TypeDBStub asyncStub() {
+        return asyncStub;
+    }
 
-        public CredentialEmbedder(String username, String password) {
-            this.username = username;
-            this.password = password;
+    @Override
+    protected <RES> RES resilientCall(Supplier<RES> function) {
+        try {
+            ensureConnected();
+            return function.get();
+        } catch (StatusRuntimeException e1) {
+            TypeDBClientException e2 = TypeDBClientException.of(e1);
+            if (e2.getErrorMessage() != null && e2.getErrorMessage().equals(ErrorMessage.Client.CLUSTER_TOKEN_CREDENTIAL_INVALID)) {
+                token = null;
+                ClusterUserTokenProto.ClusterUserToken.Renew.Res res = clusterBlockingStub.userTokenRenew(renewReq(credential.username()));
+                token = res.getToken();
+                try {
+                    return function.get();
+                } catch (StatusRuntimeException e3) {
+                    throw TypeDBClientException.of(e3);
+                }
+            } else throw e2;
         }
+    }
+
+    private class CredentialEmbedder extends CallCredentials {
```

Comment:
How about `TypeDBCallCredentials`? 
```suggestion
    private class TypeDBCallCredentials extends CallCredentials {
```

Reply:
I went for a name that's more self-explanatory since the terminology "call credential" is just not a very clear name.

The name "credential embedder" clearly indicates that the class is responsible for embedding the credential in the request.

What do you think?

Reply:
Hmm, I'm not sure that credential embedder is the best term here. The key thing is that it extends `CallCredentials`, in other words it _is a_ `CallCredentials`. Moreover, the usage of it is:
```java
this.blockingStub = TypeDBGrpc.newBlockingStub(channel).withCallCredentials(credentialEmbedder);
```
Lastly we should ask: is gRPC's name misleading? Here is the javadoc for `CallCredentials`:

https://grpc.github.io/grpc-java/javadoc/io/grpc/CallCredentials.html

The purpose of the class is to "carry credential data that will be propagated to the server via request data".

It is embedded in the request, yes, but that describes what it is used for, not what it is